src/tools/adig.c

```diff
@@ -138,14 +145,77 @@ static void print_help(void)
 }
 
 static ares_bool_t read_cmdline(int argc, const char * const *argv,
-                                adig_config_t *config)
+                                adig_config_t *config, ares_bool_t is_rcfile)
 {
-  ares_getopt_state_t state;
-  int                 c;
+  ares_getopt_state_t  state;
+  int                  c;
+  char                 configdir[256];
+  char                *configdir_xdg;
+  unsigned int         cdlen = 0;
+  unsigned int         cdsize = sizeof(configdir);
+  char                *homedir;
 
   ares_getopt_init(&state, argc, argv);
   state.opterr = 0;
 
+  if (is_rcfile) {
+    state.optind = 0;
+  } else {
+
+#if defined(WIN32)
+    cdlen = (unsigned int) snprintf(configdir, cdsize, "%s/%s", getenv("APPDATA"), "c-ares");
+
+#elif defined(__APPLE__)
+    homedir = getenv("HOME");
+    if (homedir != NULL) {
+      cdlen = (unsigned int) snprintf(configdir, cdsize, "%s/%s/%s/%s", homedir, "Library", "Application Support", "c-ares");
+    }
+
+#else
+    configdir_xdg = getenv("XDG_CONFIG_HOME");
+
+    if (configdir_xdg == NULL) {
+      homedir = getenv("HOME");
+      if (homedir != NULL) {
+        cdlen = (unsigned int) snprintf(configdir, cdsize, "%s/%s", homedir, ".config");
+      }
+    }
+    else {
+      cdlen = (unsigned int) snprintf(configdir, cdsize, "%s", configdir_xdg);
+    }
+
+#endif
+
+    DEBUGF(fprintf(stderr, "read_cmdline() configdir: %s\n", configdir));
+
+    if (cdlen != 0 && cdlen < cdsize) {
+      char            rcfile[256];
+
+      unsigned int    rclen;
+      unsigned int    rcsize = sizeof(rcfile);
+
+      size_t          rcargc;
+      char          **rcargv;
+      ares__buf_t    *rcbuf;
+
+      rclen = (unsigned int) snprintf(rcfile, rcsize, "%s/adigrc", configdir);
+      if (rclen < rcsize) {
+        rcbuf = ares__buf_create();
+        if (ares__buf_load_file(rcfile, rcbuf) == ARES_SUCCESS) {
+          ares__buf_split_str(rcbuf, (const unsigned char *)"\n ", 2, ARES_BUF_SPLIT_TRIM, 0, &rcargv, &rcargc);
```

Comment:
Should probably check this result so you don't silently ignore errors.  This will fail on things like non-printable data.

Reply:
Good point, I have it pass the error message from ares__buf_split_str() now.